File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number):
    """ A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """
    if not isinstance(number, int) or number < 1:
        raise ValueError("Classification is only possible for positive integers.")
    
    aliquot_sum = 0
    for i in range(1, number):
        if number % i == 0:
            aliquot_sum += i
    
    if aliquot_sum == number:
        return "perfect"
    elif aliquot_sum > number:
        return "abundant"
    else:
        return "deficient"

File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
//
// This is only a SKELETON file for the 'Perfect Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const classify = (number) => {
  if (number <= 0) {
    throw new Error('Classification is only possible for natural numbers.');
  }

  const aliquotSum = getAliquotSum(number);
  
  if (aliquotSum === number) {
    return 'perfect';
  } else if (aliquotSum > number) {
    return 'abundant';
  } else {
    return 'deficient';
  }
};

const getAliquotSum = (number) => {
  if (number === 1) {
    return 0;
  }

  let sum = 1;

  for (let i = 2; i * i <= number; i++) {
    if (number % i === 0) {
      sum += i;

      if (i * i !== number) {
        sum += number / i;
      }
    }
  }
  return sum;
}